accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,608 @@
+
+## Primary API
+All types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
+{
+    public abstract class JsonNode
+    {
+        public JsonNode(JsonSerializerOptions options = null);
+
+        public JsonSerializerOptions Options { get; }
+
+        // Return the internal value or convert to T as necessary.
+        public abstract T GetValue<T>();
+        public abstract object GetValue<Type type>();
+        public abstract bool TryGetValue<T>(out T? value);
+        public abstract bool TryGetValue(Type type, out object? value);
+
+        // The token type from deserialization; otherwise JsonValueKind.Unspecified.
+        public JsonValueKind ValueKind { get; }
+    }
+
+    public sealed class JsonArray : JsonNode, IList<object>
+    {
+        public JsonArray(JsonSerializerOptions options = null);
+        public JsonArray(JsonElement value, JsonSerializerOptions options = null);
+
+        // Can be used to return:
+        // - Any compatible collection type
+        // - JsonElement
+        // - string or byte[] (as JSON)
+        public override T GetValue<T>();
+        public override object GetValue<Type type>();
+        public override bool TryGetValue<T>(out T? value);
+        public override bool TryGetValue(Type type, out object? value);
+    }
+
+    public sealed class JsonObject : JsonNode, IDictionary<string, object>
+    {
+        public JsonObject(JsonSerializerOptions options = null);
+        public JsonObject(JsonElement value, JsonSerializerOptions options = null);
+
+        // Can be used to return:
+        // - Any compatible dictionary type
+        // - POCO including anonymous type or record
+        // - JsonElement
+        // - string or byte[] (as JSON)
+        public override T GetValue<T>();
+        public override object GetValue<Type type>();
+        public override bool TryGetValue<T>(out T? value);
+        public override bool TryGetValue(Type type, out object? value);
+    }
+
+    public sealed class JsonValue : JsonNode
+    {
+        public JsonValue(JsonSerializerOptions options = null);
+        public JsonValue(JsonElement value, JsonSerializerOptions options = null);
+        public JsonValue(object value, JsonSerializerOptions options = null);
+
+        // The internal raw value.
+        // During deserialization, this will be a JsonElement (pending design) since it supports
+        // delayed creation of string and on-demand creation of numbers.
+        // In edit mode, can be set to any object.
+        public object Value {get; set;}
```

Comment:
Note that the generic versions below won't box; not sure on why we wouldn't want `object` here since we want to allow any CLR value to be assigned.

Reply:
With the current design, not sure how this could not be `object`. i.e. if `JsonNode` then we wouldn't be able to assign any CLR object, unless the other design used generics for the internal value?

Reply:
I agree, we should avoid such hidden boxing for primitive JSON value tokens.